File: python-package/src/aabatlas/datasets.py
"""List and resolve atlas records."""
from __future__ import annotations

from typing import Any

import pandas as pd

from ._source import fetch


def list_datasets(ref: str = "main") -> pd.DataFrame:
    """List all datasets/papers in the atlas.

    Columns match ``schema/dataset.schema.json`` in the atlas repo
    (``atlas_id``, ``record_type``, ``title``, ``organism``, ``conditions``,
    ``total_sample_n``, ``platforms``, ``dataset_link``,
    ``curated_availability``, etc.). List/dict-valued fields come back as
    Python objects in object-dtype columns.
    """
    return fetch("datasets", ref=ref)


def list_publications(ref: str = "main") -> pd.DataFrame:
    """List all publications referenced by atlas datasets."""
    return fetch("publications", ref=ref)


def list_platforms(ref: str = "main") -> pd.DataFrame:
    """List all assay platforms referenced by atlas datasets.

    Per-study protein/feature counts live on the dataset record
    (``list_datasets()["platforms"]``), not here -- see ``schema/README.md``
    in the atlas repo for why.
    """
    return fetch("platforms", ref=ref)


def list_link_status(ref: str = "main") -> pd.DataFrame:
    """List the link-reachability check history.

    One row per automated check, appended over time by the atlas's
    scheduled link checker. ``reachable`` is the raw HTTP signal;
    ``likely_bot_blocked`` flags 403/429 responses from publisher anti-bot
    defenses, which are not necessarily dead links.
    """
    return fetch("link_status", ref=ref)


def get_dataset(atlas_id: str, ref: str = "main") -> dict[str, Any]:
    """Get a single dataset record, with its publication and platforms resolved.

    Returns a dict with the dataset's fields plus two resolved keys:
    ``publication`` (a dict, or ``None`` if the record has no associated
    publication) and ``platforms_detail`` (a list combining each
    ``platforms[].platform_id`` from the dataset with the matching row from
    ``list_platforms()``, keeping ``n_proteins_reported``).
    """
    ds = list_datasets(ref=ref)
    match = ds[ds["atlas_id"] == atlas_id]
    if match.empty:
        raise ValueError(f"No dataset found with atlas_id: {atlas_id}")
    record = match.iloc[0].to_dict()

    pub_id = record.get("publication_id")
    if pub_id:
        pubs = list_publications(ref=ref)
        pub_match = pubs[pubs["publication_id"] == pub_id]
        record["publication"] = pub_match.iloc[0].to_dict() if not pub_match.empty else None
    else:
        record["publication"] = None

    platforms = record.get("platforms") or []
    detail = []
    if platforms:
        all_platforms = list_platforms(ref=ref)
        for p in platforms:
            merged = dict(p)
            prow = all_platforms[all_platforms["platform_id"] == p["platform_id"]]
            if not prow.empty:
                merged.update({k: v for k, v in prow.iloc[0].to_dict().items() if k != "platform_id"})
            detail.append(merged)
    record["platforms_detail"] = detail

    return record

File: r-package/R/datasets.R
#' List all datasets/papers in the atlas
#'
#' @inheritParams aab_raw_url
#' @return A [tibble::tibble()], one row per record, with columns matching
#'   `schema/dataset.schema.json` (`atlas_id`, `record_type`, `title`,
#'   `organism`, `conditions`, `total_sample_n`, `platforms`, `dataset_link`,
#'   `curated_availability`, etc.). List-valued fields come back as
#'   list-columns.
#' @examples
#' \dontrun{
#' ds <- aab_datasets()
#' ds[ds$record_type == "DATASET", c("atlas_id", "title")]
#' }
#' @export
aab_datasets <- function(ref = "main") {
  aab_fetch("datasets", ref = ref)
}

#' List all publications referenced by atlas datasets
#'
#' @inheritParams aab_raw_url
#' @return A [tibble::tibble()] with columns `publication_id`, `citation`,
#'   `doi`, `pubmed_link`, `publisher_link`.
#' @export
aab_publications <- function(ref = "main") {
  aab_fetch("publications", ref = ref)
}

#' List all assay platforms referenced by atlas datasets
#'
#' @inheritParams aab_raw_url
#' @return A [tibble::tibble()] with columns `platform_id`, `name`,
#'   `technology`. Per-study protein/feature counts live on
#'   the dataset record (`aab_datasets()$platforms`), not here — see
#'   `schema/README.md` in the atlas repo for why.
#' @export
aab_platforms <- function(ref = "main") {
  aab_fetch("platforms", ref = ref)
}

#' List the link-reachability check history
#'
#' @inheritParams aab_raw_url
#' @return A [tibble::tibble()] with one row per automated check, appended
#'   over time by the atlas's scheduled link checker — see
#'   `checked_at`/`reachable` to see whether a given dataset's link has gone
#'   stale.
#' @export
aab_link_status <- function(ref = "main") {
  aab_fetch("link_status", ref = ref)
}

#' Get a single dataset record, with its publication and platforms resolved
#'
#' @param atlas_id An atlas id, e.g. `"AAB-000004"`.
#' @inheritParams aab_raw_url
#' @return A list with the dataset's fields plus two resolved fields:
#'   `publication` (a one-row tibble, or `NULL` if none) and
#'   `platforms_detail` (a tibble joining `platforms[].platform_id` from the
#'   dataset onto `aab_platforms()`, keeping `n_proteins_reported`).
#' @examples
#' \dontrun{
#' aab_dataset("AAB-000004")
#' }
#' @export
aab_dataset <- function(atlas_id, ref = "main") {
  datasets <- aab_datasets(ref = ref)
  row <- datasets[datasets$atlas_id == atlas_id, ]
  if (nrow(row) == 0) {
    stop("No dataset found with atlas_id: ", atlas_id, call. = FALSE)
  }
  record <- as.list(row[1, ])

  if (!is.null(record$publication_id) && !is.na(record$publication_id)) {
    pubs <- aab_publications(ref = ref)
    record$publication <- pubs[pubs$publication_id == record$publication_id, ]
  } else {
    record$publication <- NULL
  }

  plats <- record$platforms[[1]]
  if (!is.null(plats) && nrow(plats) > 0) {
    all_platforms <- aab_platforms(ref = ref)
    record$platforms_detail <- merge(plats, all_platforms, by = "platform_id", all.x = TRUE)
  } else {
    record$platforms_detail <- NULL
  }

  record
}
